File: 015.three_sum/solution.rb
require 'set'

# @param {Integer[]} nums
# @return {Integer[][]}
def three_sum(nums)
    nums.sort!
    results = []
    i = 0
    n = nums.size
    while i < n && nums[i] <= 0 do
        two_sum(nums, i, results) if i == 0 || nums[i-1] != nums[i] 
        i += 1
    end
    results
end

def two_sum(nums, i, results)
    lo = i+1
    hi = nums.size - 1
    while lo < hi do
        sum = nums[i] + nums[lo] + nums[hi]
        if sum < 0
            lo += 1
        elsif sum > 0
            hi -= 1
        else
            results.push([nums[i], nums[lo], nums[hi]])
            lo += 1
            hi -= 1
            while lo < hi && nums[lo] == nums[lo-1] do
                lo += 1
            end
        end
    end
end

File: 015.three_sum/solution.ex
defmodule Solution do
  @spec three_sum(nums :: [integer]) :: [[integer]]
  def three_sum(nums) do
    xs = nums |> Enum.sort() |> Enum.with_index()
    three_sum_list(xs, Enumm.reverse(xs))
  end

  def three_sum_list([], rev_list), do: []
  def three_sum_list([{x, _}| _], _) when x > 0, do: []
  def three_sum_list([{x, i}| xis], rev) do
    List.flatten(two_sum_list(x, i, xis, rev), three_sum_list(xis, rev))
  end

  def two_sum_list(_, _, [], _), do: []
  def two_sum_list(_, _, _, []), do: []
  def two_sum_list(x, i, [{y, y_i} | yis], [{z, z_i} | zis]) when y_i >= z_i, do: []
  def two_sum_list(x, i, [{y, y_i} | y_next] = yis, [{z, z_i} | z_next] = zis) do
    cond do
      x + y + z < 0 -> two_sum_list(x, i, y_next, zis)
      x + y + z > 0 -> two_sum_list(x, i, yis, z_next)
      true -> [{x, y, z} | two_sum_list(x, i, Enum.drop_while(y_next, &(&1 == y)), z_next)]
    end
  end
end
